libcasr/src/sarif.rs

```diff
@@ -0,0 +1,404 @@
+//! Sarif module contains `Sarif` struct that countains multiple `CrashReport`
+//! structs in SARIF format.
+
+use crate::constants::*;
+use crate::error::{Error, Result};
+use crate::execution_class::{ExecutionClass, CLASSES};
+use crate::init_ignored_frames;
+use crate::report::CrashReport;
+use crate::stacktrace::{STACK_FRAME_FILEPATH_IGNORE_REGEXES, STACK_FRAME_FUNCTION_IGNORE_REGEXES};
+
+use serde_json::{Map, Value};
+
+use std::path::Path;
+
+/// CASR CrashReports in SARIF format.
+#[derive(Clone, Debug, Default)]
+pub struct SarifReport {
+    /// SARIF json.
+    pub json: Value,
+    /// current rule id for generated ExecutionClasses.
+    id: u64,
+}
+
+impl SarifReport {
+    /// Create new `SarifReport` with blank required fields
+    pub fn new() -> Self {
+        let mut map = Map::new();
+        map.insert(
+            "$schema".to_string(),
+            Value::String("https://json.schemastore.org/sarif-2.1.0.json".to_string()),
+        );
+        map.insert("version".to_string(), Value::String("2.1.0".to_string()));
+        let mut runs = Vec::new();
+        let mut run = Map::new();
+        let mut tool = Map::new();
+        let mut driver = Map::new();
+        driver.insert("name".to_string(), Value::String("CASR".to_string()));
+        driver.insert("rules".to_string(), Value::Array(Vec::new()));
+        tool.insert("driver".to_string(), Value::Object(driver));
+        run.insert("tool".to_string(), Value::Object(tool));
+        run.insert("results".to_string(), Value::Array(Vec::new()));
+        runs.push(Value::Object(run));
+        map.insert("runs".to_string(), Value::Array(runs));
+        Self {
+            json: Value::Object(map),
+            id: 0,
+        }
+    }
+
+    /// Set name for SARIF tool:driver.
+    /// NOTE: before use this method,
+    /// use SarifReprot::new() to get report.
+    ///
+    ///  # Arguments
+    ///
+    ///  * 'name' - tool:driver name (default is CASR)
+    pub fn set_name(&mut self, name: &str) {
+        let driver = self.json.as_object_mut().unwrap()["runs"]
+            .as_array_mut()
+            .unwrap()[0]
+            .as_object_mut()
+            .unwrap()["tool"]
+            .as_object_mut()
+            .unwrap()["driver"]
+            .as_object_mut()
+            .unwrap();
+        driver.insert("name".to_string(), Value::String(name.to_string()));
+    }
+
+    /// SARIF rule from ExecutionClass.
+    /// NOTE: before using this method,
+    /// use SarifReprot::new() to get report.
+    ///
+    ///  # Arguments
+    ///
+    ///  * 'class' - ExecutionClass from CrashReport
+    ///
+    ///  # Return
+    ///
+    ///  If Rule is new, than it's json object is returned and ruleId
+    ///  else ruleId only is returned.
+    fn rule(&self, class: &ExecutionClass) -> (Option<Value>, String) {
+        let rule_id = if let Some(pos) = CLASSES
+            .iter()
+            .position(|item| item.1 == class.short_description)
```

Comment:
what if we change elements order in CLASSES vector? Maybe we can store ids there or add a comment that new classes should always be added to the end? And what if we remove a class from CLASSES?

Reply:
Maybe we need a comment and think about this case.